File: picasso/src/test/java/com/squareup/picasso3/NetworkRequestHandlerTest.java
/*
 * Copyright (C) 2013 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.picasso3;

import android.net.NetworkInfo;
import android.net.Uri;
import android.support.annotation.NonNull;
import com.squareup.picasso3.RequestHandler.Result;
import com.squareup.picasso3.TestUtils.PremadeCall;
import java.util.concurrent.BlockingDeque;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.atomic.AtomicBoolean;
import okhttp3.CacheControl;
import okhttp3.Call;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.robolectric.RobolectricTestRunner;

import static com.google.common.truth.Truth.assertThat;
import static com.squareup.picasso3.TestUtils.URI_1;
import static com.squareup.picasso3.TestUtils.URI_KEY_1;
import static com.squareup.picasso3.TestUtils.mockNetworkInfo;
import static java.util.concurrent.TimeUnit.SECONDS;
import static okhttp3.Protocol.HTTP_1_1;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

@RunWith(RobolectricTestRunner.class)
public class NetworkRequestHandlerTest {
  final BlockingDeque<Response> responses = new LinkedBlockingDeque<>();
  final BlockingDeque<okhttp3.Request> requests = new LinkedBlockingDeque<>();

  @Mock Picasso picasso;
  @Mock Stats stats;
  @Mock Dispatcher dispatcher;

  private NetworkRequestHandler networkHandler;

  @Before public void setUp() {
    initMocks(this);
    networkHandler = new NetworkRequestHandler(new Call.Factory() {
      @Override public Call newCall(Request request) {
        requests.add(request);
        try {
          return new PremadeCall(request, responses.takeFirst());
        } catch (InterruptedException e) {
          throw new AssertionError(e);
        }
      }
    }, stats);
  }

  @Test public void doesNotForceLocalCacheOnlyWithAirplaneModeOffAndRetryCount() throws Exception {
    responses.add(responseOf(ResponseBody.create(null, new byte[10])));
    Action action = TestUtils.mockAction(URI_KEY_1, URI_1);
    final CountDownLatch latch = new CountDownLatch(1);
    networkHandler.load(picasso, action.request, new RequestHandler.Callback() {
      @Override public void onSuccess(Result result) {
        try {
          assertThat(requests.takeFirst().cacheControl().toString()).isEmpty();
          latch.countDown();
        } catch (InterruptedException e) {
          throw new AssertionError(e);
        }
      }

      @Override public void onError(@NonNull Throwable t) {
        throw new AssertionError(t);
      }
    });
    assertThat(latch.await(10, SECONDS)).isTrue();
  }

  @Test public void withZeroRetryCountForcesLocalCacheOnly() throws Exception {
    responses.add(responseOf(ResponseBody.create(null, new byte[10])));
    Action action = TestUtils.mockAction(URI_KEY_1, URI_1);
    PlatformLruCache cache = new PlatformLruCache(0);
    BitmapHunter hunter =
        new BitmapHunter(picasso, dispatcher, cache, stats, action, networkHandler);
    hunter.retryCount = 0;
    hunter.hunt();
    assertThat(requests.takeFirst().cacheControl().toString()).isEqualTo(CacheControl.FORCE_CACHE.toString());
  }

  @Test public void shouldRetryTwiceWithAirplaneModeOffAndNoNetworkInfo() {
    Action action = TestUtils.mockAction(URI_KEY_1, URI_1);
    PlatformLruCache cache = new PlatformLruCache(0);
    BitmapHunter hunter =
        new BitmapHunter(picasso, dispatcher, cache, stats, action, networkHandler);
    assertThat(hunter.shouldRetry(false, null)).isTrue();
    assertThat(hunter.shouldRetry(false, null)).isTrue();
    assertThat(hunter.shouldRetry(false, null)).isFalse();
  }

  @Test public void shouldRetryWithUnknownNetworkInfo() {
    assertThat(networkHandler.shouldRetry(false, null)).isTrue();
    assertThat(networkHandler.shouldRetry(true, null)).isTrue();
  }

  @Test public void shouldRetryWithConnectedNetworkInfo() {
    NetworkInfo info = mockNetworkInfo();
    when(info.isConnected()).thenReturn(true);
    assertThat(networkHandler.shouldRetry(false, info)).isTrue();
    assertThat(networkHandler.shouldRetry(true, info)).isTrue();
  }

  @Test public void shouldNotRetryWithDisconnectedNetworkInfo() {
    NetworkInfo info = mockNetworkInfo();
    when(info.isConnectedOrConnecting()).thenReturn(false);
    assertThat(networkHandler.shouldRetry(false, info)).isFalse();
    assertThat(networkHandler.shouldRetry(true, info)).isFalse();
  }

  @Test public void noCacheAndKnownContentLengthDispatchToStats() throws Exception {
    responses.add(responseOf(ResponseBody.create(null, new byte[10])));
    Action action = TestUtils.mockAction(URI_KEY_1, URI_1);
    final CountDownLatch latch = new CountDownLatch(1);
    networkHandler.load(picasso, action.request, new RequestHandler.Callback() {
      @Override public void onSuccess(Result result) {
        verify(stats).dispatchDownloadFinished(10);
        latch.countDown();
      }

      @Override public void onError(@NonNull Throwable t) {
        throw new AssertionError(t);
      }
    });
    assertThat(latch.await(10, SECONDS)).isTrue();
  }

  @Test public void unknownContentLengthFromDiskThrows() throws Exception {
    final AtomicBoolean closed = new AtomicBoolean();
    ResponseBody body = new ResponseBody() {
      @Override public MediaType contentType() { return null; }
      @Override public long contentLength() { return 0; }
      @Override public BufferedSource source() { return new Buffer(); }
      @Override public void close() {
        closed.set(true);
        super.close();
      }
    };
    responses.add(responseOf(body)
        .newBuilder()
        .cacheResponse(responseOf(null))
        .build());
    Action action = TestUtils.mockAction(URI_KEY_1, URI_1);
    final CountDownLatch latch = new CountDownLatch(1);
    networkHandler.load(picasso, action.request, new RequestHandler.Callback() {
      @Override public void onSuccess(Result result) {
        throw new AssertionError();
      }

      @Override public void onError(@NonNull Throwable t) {
        verifyZeroInteractions(stats);
        assertTrue(closed.get());
        latch.countDown();
      }
    });
    assertThat(latch.await(10, SECONDS)).isTrue();
  }

  @Test public void cachedResponseDoesNotDispatchToStats() throws Exception {
    responses.add(responseOf(ResponseBody.create(null, new byte[10]))
        .newBuilder()
        .cacheResponse(responseOf(null))
        .build());
    Action action = TestUtils.mockAction(URI_KEY_1, URI_1);
    final CountDownLatch latch = new CountDownLatch(1);
    networkHandler.load(picasso, action.request, new RequestHandler.Callback() {
      @Override public void onSuccess(Result result) {
        verifyZeroInteractions(stats);
        latch.countDown();
      }

      @Override public void onError(@NonNull Throwable t) {
        throw new AssertionError(t);
      }
    });
    assertThat(latch.await(10, SECONDS)).isTrue();
  }

  @Test public void shouldHandleSchemeInsensitiveCase() {
    String[] schemes = {
            "http",
            "https",
            "HTTP",
            "HTTPS",
            "HTtP",
    };

    for (String scheme : schemes) {
      Uri uri = URI_1.buildUpon().scheme(scheme).build();
      assertThat(networkHandler.canHandleRequest(TestUtils.mockRequest(uri))).isTrue();
    }
  }


  private static Response responseOf(ResponseBody body) {
    return new Response.Builder()
        .code(200)
        .protocol(HTTP_1_1)
        .request(new okhttp3.Request.Builder().url("http://example.com").build())
        .message("OK")
        .body(body)
        .build();
  }
}

File: picasso/src/test/java/com/squareup/picasso3/NetworkRequestHandlerTest.kt
/*
 * Copyright (C) 2013 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.picasso3

import com.google.common.truth.Truth.assertThat
import com.squareup.picasso3.RequestHandler.Result
import com.squareup.picasso3.TestUtils.CUSTOM_HEADER_NAME
import com.squareup.picasso3.TestUtils.CUSTOM_HEADER_VALUE
import com.squareup.picasso3.TestUtils.EventRecorder
import com.squareup.picasso3.TestUtils.PremadeCall
import com.squareup.picasso3.TestUtils.URI_1
import com.squareup.picasso3.TestUtils.URI_KEY_1
import com.squareup.picasso3.TestUtils.mockNetworkInfo
import com.squareup.picasso3.TestUtils.mockPicasso
import okhttp3.CacheControl
import okhttp3.MediaType
import okhttp3.Protocol.HTTP_1_1
import okhttp3.Response
import okhttp3.ResponseBody
import okhttp3.ResponseBody.Companion.toResponseBody
import okio.Buffer
import okio.BufferedSource
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.MockitoAnnotations.initMocks
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import java.util.concurrent.CountDownLatch
import java.util.concurrent.LinkedBlockingDeque
import java.util.concurrent.TimeUnit.SECONDS
import java.util.concurrent.atomic.AtomicBoolean

@RunWith(RobolectricTestRunner::class)
class NetworkRequestHandlerTest {
  private val responses = LinkedBlockingDeque<Response>()
  private val requests = LinkedBlockingDeque<okhttp3.Request>()

  @Mock internal lateinit var dispatcher: Dispatcher
  private lateinit var picasso: Picasso
  private lateinit var networkHandler: NetworkRequestHandler

  @Before fun setUp() {
    initMocks(this)
    picasso = mockPicasso(RuntimeEnvironment.application)
    networkHandler = NetworkRequestHandler { request ->
      requests.add(request)
      try {
        PremadeCall(request, responses.takeFirst())
      } catch (e: InterruptedException) {
        throw AssertionError(e)
      }
    }
  }

  @Test fun doesNotForceLocalCacheOnlyWithAirplaneModeOffAndRetryCount() {
    responses.add(responseOf(ByteArray(10).toResponseBody(null)))
    val action = TestUtils.mockAction(picasso, URI_KEY_1, URI_1)
    val latch = CountDownLatch(1)
    networkHandler.load(
      picasso = picasso,
      request = action.request,
      callback = object : RequestHandler.Callback {
        override fun onSuccess(result: Result?) {
          try {
            assertThat(requests.takeFirst().cacheControl.toString()).isEmpty()
            latch.countDown()
          } catch (e: InterruptedException) {
            throw AssertionError(e)
          }
        }

        override fun onError(t: Throwable): Unit = throw AssertionError(t)
      }
    )
    assertThat(latch.await(10, SECONDS)).isTrue()
  }

  @Test fun withZeroRetryCountForcesLocalCacheOnly() {
    responses.add(responseOf(ByteArray(10).toResponseBody(null)))
    val action = TestUtils.mockAction(picasso, URI_KEY_1, URI_1)
    val cache = PlatformLruCache(0)
    val hunter = BitmapHunter(picasso, dispatcher, cache, action, networkHandler)
    hunter.retryCount = 0
    hunter.hunt()
    assertThat(requests.takeFirst().cacheControl.toString())
      .isEqualTo(CacheControl.FORCE_CACHE.toString())
  }

  @Test fun shouldRetryTwiceWithAirplaneModeOffAndNoNetworkInfo() {
    val action = TestUtils.mockAction(picasso, URI_KEY_1, URI_1)
    val cache = PlatformLruCache(0)
    val hunter = BitmapHunter(picasso, dispatcher, cache, action, networkHandler)
    assertThat(hunter.shouldRetry(airplaneMode = false, info = null)).isTrue()
    assertThat(hunter.shouldRetry(airplaneMode = false, info = null)).isTrue()
    assertThat(hunter.shouldRetry(airplaneMode = false, info = null)).isFalse()
  }

  @Test fun shouldRetryWithUnknownNetworkInfo() {
    assertThat(networkHandler.shouldRetry(airplaneMode = false, info = null)).isTrue()
    assertThat(networkHandler.shouldRetry(airplaneMode = true, info = null)).isTrue()
  }

  @Test fun shouldRetryWithConnectedNetworkInfo() {
    val info = mockNetworkInfo()
    `when`(info.isConnected).thenReturn(true)
    assertThat(networkHandler.shouldRetry(airplaneMode = false, info = info)).isTrue()
    assertThat(networkHandler.shouldRetry(airplaneMode = true, info = info)).isTrue()
  }

  @Test fun shouldNotRetryWithDisconnectedNetworkInfo() {
    val info = mockNetworkInfo()
    `when`(info.isConnectedOrConnecting).thenReturn(false)
    assertThat(networkHandler.shouldRetry(airplaneMode = false, info = info)).isFalse()
    assertThat(networkHandler.shouldRetry(airplaneMode = true, info = info)).isFalse()
  }

  @Test fun noCacheAndKnownContentLengthDispatchToStats() {
    val eventRecorder = EventRecorder()
    val picasso = picasso.newBuilder().addEventListener(eventRecorder).build()
    val knownContentLengthSize = 10
    responses.add(responseOf(ByteArray(knownContentLengthSize).toResponseBody(null)))
    val action = TestUtils.mockAction(picasso, URI_KEY_1, URI_1)
    val latch = CountDownLatch(1)
    networkHandler.load(
      picasso = picasso,
      request = action.request,
      callback = object : RequestHandler.Callback {
        override fun onSuccess(result: Result?) {
          assertThat(eventRecorder.downloadSize).isEqualTo(knownContentLengthSize)
          latch.countDown()
        }

        override fun onError(t: Throwable): Unit = throw AssertionError(t)
      }
    )
    assertThat(latch.await(10, SECONDS)).isTrue()
  }

  @Test fun unknownContentLengthFromDiskThrows() {
    val eventRecorder = EventRecorder()
    val picasso = picasso.newBuilder().addEventListener(eventRecorder).build()
    val closed = AtomicBoolean()
    val body = object : ResponseBody() {
      override fun contentType(): MediaType? = null
      override fun contentLength(): Long = 0
      override fun source(): BufferedSource = Buffer()
      override fun close() {
        closed.set(true)
        super.close()
      }
    }
    responses += responseOf(body)
      .newBuilder()
      .cacheResponse(responseOf(null))
      .build()
    val action = TestUtils.mockAction(picasso, URI_KEY_1, URI_1)
    val latch = CountDownLatch(1)
    networkHandler.load(
      picasso = picasso,
      request = action.request,
      callback = object : RequestHandler.Callback {
        override fun onSuccess(result: Result?): Unit = throw AssertionError()

        override fun onError(t: Throwable) {
          assertThat(eventRecorder.downloadSize).isEqualTo(0)
          assertTrue(closed.get())
          latch.countDown()
        }
      }
    )
    assertThat(latch.await(10, SECONDS)).isTrue()
  }

  @Test fun cachedResponseDoesNotDispatchToStats() {
    val eventRecorder = EventRecorder()
    val picasso = picasso.newBuilder().addEventListener(eventRecorder).build()
    responses += responseOf(ByteArray(10).toResponseBody(null))
      .newBuilder()
      .cacheResponse(responseOf(null))
      .build()
    val action = TestUtils.mockAction(picasso, URI_KEY_1, URI_1)
    val latch = CountDownLatch(1)
    networkHandler.load(
      picasso = picasso,
      request = action.request,
      callback = object : RequestHandler.Callback {
        override fun onSuccess(result: Result?) {
          assertThat(eventRecorder.downloadSize).isEqualTo(0)
          latch.countDown()
        }

        override fun onError(t: Throwable): Unit = throw AssertionError(t)
      }
    )
    assertThat(latch.await(10, SECONDS)).isTrue()
  }

  @Test fun customHeaders() {
    responses += responseOf(ByteArray(10).toResponseBody(null))
      .newBuilder()
      .cacheResponse(responseOf(null))
      .build()
    val action = TestUtils.mockAction(
      picasso,
      key = URI_KEY_1,
      uri = URI_1,
      headers = mapOf(CUSTOM_HEADER_NAME to CUSTOM_HEADER_VALUE)
    )
    val latch = CountDownLatch(1)
    networkHandler.load(
      picasso = picasso,
      request = action.request,
      callback = object : RequestHandler.Callback {
        override fun onSuccess(result: Result?) {
          with(requests.first.headers) {
            assertThat(names()).containsExactly(CUSTOM_HEADER_NAME)
            assertThat(values(CUSTOM_HEADER_NAME)).containsExactly(CUSTOM_HEADER_VALUE)
          }
          latch.countDown()
        }

        override fun onError(t: Throwable): Unit = throw AssertionError(t)
      }
    )
    assertThat(latch.await(10, SECONDS)).isTrue()
  }

  @Test fun shouldHandleSchemeInsensitiveCase() {
    val schemes = arrayOf("http", "https", "HTTP", "HTTPS", "HTtP")
    for (scheme in schemes) {
      val uri = URI_1.buildUpon().scheme(scheme).build()
      assertThat(networkHandler.canHandleRequest(TestUtils.mockRequest(uri))).isTrue()
    }
  }

  private fun responseOf(body: ResponseBody?) =
    Response.Builder()
      .code(200)
      .protocol(HTTP_1_1)
      .request(okhttp3.Request.Builder().url("http://example.com").build())
      .message("OK")
      .body(body)
      .build()
}
